pandera/engines/polars_engine.py

```diff
@@ -0,0 +1,496 @@
+"""Polars engine and data types."""
+import dataclasses
+import datetime
+import decimal
+import inspect
+import warnings
+from typing import Any, Union, Optional, Iterable, Literal
+
+
+import polars as pl
+from polars.datatypes import py_type_to_dtype
+
+from pandera import dtypes, errors
+from pandera.dtypes import immutable
+from pandera.engines import engine
+from pandera.engines.type_aliases import PolarsObject
+from pandera.engines.utils import (
+    polars_coerce_failure_cases,
+    polars_object_coercible,
+    polars_failure_cases_from_coercible,
+    check_polars_container_all_true,
+)
+
+
+@immutable(init=True)
+class DataType(dtypes.DataType):
+    """Base `DataType` for boxing Polars data types."""
+
+    type: pl.DataType = dataclasses.field(repr=False, init=False)
+
+    def __init__(self, dtype: Optional[Any] = None):
+        super().__init__()
+        try:
+            object.__setattr__(self, "type", py_type_to_dtype(dtype))
+        except ValueError:
+            object.__setattr__(self, "type", pl.Object)
+
+        dtype_cls = dtype if inspect.isclass(dtype) else dtype.__class__
+        warnings.warn(
+            f"'{dtype_cls}' support is not guaranteed.\n"
+            + "Usage Tip: Consider writing a custom "
+            + "pandera.dtypes.DataType or opening an issue at "
+            + "https://github.com/pandera-dev/pandera"
+        )
+
+    def __post_init__(self):
+        # this method isn't called if __init__ is defined
+        object.__setattr__(
+            self, "type", py_type_to_dtype(self.type)
+        )  # pragma: no cover
+
+    def coerce(self, data_container: PolarsObject) -> PolarsObject:
+        """Coerce data container to the data type."""
+        return data_container.cast(self.type, strict=True)
+
+    def try_coerce(self, data_container: PolarsObject) -> PolarsObject:
+        """Coerce data container to the data type,
+        raises a :class:`~pandera.errors.ParserError` if the coercion fails
+        :raises: :class:`~pandera.errors.ParserError`: if coercion fails
+        """
+        try:
+            return self.coerce(data_container)
+        except Exception as exc:  # pylint:disable=broad-except
+            raise errors.ParserError(
+                f"Could not coerce {type(data_container)} data_container "
+                f"into type {self.type}",
+                failure_cases=polars_coerce_failure_cases(
+                    data_container=data_container, type_=self.type
+                ),
+            ) from exc
+
+    def check(
+        self,
+        pandera_dtype: dtypes.DataType,
+        data_container: Optional[PolarsObject] = None,
+    ) -> Union[bool, Iterable[bool]]:
+        try:
+            return self.type == pandera_dtype
```

Comment:
I think this should be `self.type == pandera_dtype.type`. Related to this, maybe good to add a test like `test_check_equivalent`

Reply:
Yes you are correct, I will fix this and add a test just for this case.